heat/core/manipulations.py

```diff
@@ -93,62 +102,80 @@ def concatenate(arrays, axis=0):
     """
     if not isinstance(arrays, (tuple, list)):
         raise TypeError("arrays must be a list or a tuple")
+
+    # a single array cannot be concatenated
     if len(arrays) < 2:
         raise ValueError("concatenate requires 2 arrays")
+    # concatenate multiple arrays
     elif len(arrays) > 2:
         res = concatenate((arrays[0], arrays[1]), axis=axis)
         for a in range(2, len(arrays)):
             res = concatenate((res, arrays[a]), axis=axis)
         return res
 
+    # unpack the arrays
     arr0, arr1 = arrays[0], arrays[1]
 
+    # input sanitation
     if not isinstance(arr0, dndarray.DNDarray) or not isinstance(arr1, dndarray.DNDarray):
         raise TypeError("Both arrays must be DNDarrays")
+
     if not isinstance(axis, int):
         raise TypeError("axis must be an integer, currently: {}".format(type(axis)))
-
     axis = stride_tricks.sanitize_axis(arr0.gshape, axis)
 
     if arr0.numdims != arr1.numdims:
-        raise RuntimeError("DNDarrays must have the same number of dimensions")
+        raise ValueError("DNDarrays must have the same number of dimensions")
 
     if not all([arr0.gshape[i] == arr1.gshape[i] for i in range(len(arr0.gshape)) if i != axis]):
         raise ValueError(
-            "Arrays cannot be concatenated, gshapes must be the same in every axis "
+            "Arrays cannot be concatenated, shapes must be the same in every axis "
             "except the selected axis: {}, {}".format(arr0.gshape, arr1.gshape)
         )
 
-    s0, s1 = arr0.split, arr1.split
+    # different communicators may not be concatenated
+    if arr0.comm != arr1.comm:
+        raise RuntimeError("Communicators of passed arrays mismatch.")
```

Comment:
i think that this is a dangerous precedent to set. because when we create new arrays we frequently do not ensure that they have the same comms. we could generate one global comm and the use this for all created arrays in the script. that would ensure that we have the same object as default everywhere

Reply:
Well we do have one global communicator, which is COMM_WORLD and we actually need to do what is coded above everywhere. What happens if we have two different comms for some variables a and b that have size 3 and 5 respectively?